go-controller/pkg/ovnwebhook/podadmission.go

```diff
@@ -0,0 +1,138 @@
+		if err != nil {
+			return err
+		}
+		node, err := nodeLister.Get(nodeName)
+		if err != nil {
+			return fmt.Errorf("could not get info on node %s from client: %w", nodeName, err)
+		}
+
+		subnets, err := util.ParseNodeHostSubnetAnnotation(node, types.DefaultNetworkName)
+		if err != nil {
+			return err
+		}
+		for _, ip := range podAnnot.IPs {
+			found := false
+			for _, subnet := range subnets {
+				if subnet.Contains(ip.IP) {
+					found = true
+				}
+			}
+			if !found {
+				return fmt.Errorf("%s does not belong to %s node", ip, nodeName)
+			}
+		}
+		return nil
+	},
+	util.DPUConnectionDetailsAnnot: nil,
+	util.DPUConnetionStatusAnnot:   nil,
+}
+
+type PodAdmission struct {
+	nodeLister     listers.NodeLister
+	annotations    map[string]checkPodAnnot
+	annotationKeys sets.Set[string]
+}
+
+func NewPodAdmissionWebhook(nodeLister listers.NodeLister) *PodAdmission {
+	return &PodAdmission{
+		nodeLister:     nodeLister,
+		annotations:    interconnectPodAnnotations,
+		annotationKeys: sets.New[string](maps.Keys(interconnectPodAnnotations)...),
+	}
+}
+
+func (p PodAdmission) ValidateCreate(ctx context.Context, obj runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore creation, the webhook is configured to only handle pod/status updates
+	return nil, nil
+}
+
+func (p PodAdmission) ValidateDelete(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore creation, the webhook is configured to only handle pod/status updates
+	return nil, nil
+}
+
+var _ admission.CustomValidator = &PodAdmission{}
+
+func (p PodAdmission) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (warnings admission.Warnings, err error) {
+	oldPod := oldObj.(*corev1.Pod)
+	newPod := newObj.(*corev1.Pod)
+
+	req, err := admission.RequestFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	nodeName, isOVNKubeNode := ovnkubeNodeIdentity(req.UserInfo)
+
+	if !isOVNKubeNode {
+		// the user is not an ovnkube-node
+		return nil, nil
+	}
+
+	if oldPod.Spec.NodeName != nodeName {
+		return nil, fmt.Errorf("ovnkube-node on node: %q is not allowed to modify pods %q annotations", nodeName, oldPod.Name)
+	}
+	if newPod.Spec.NodeName != nodeName {
+		return nil, fmt.Errorf("ovnkube-node on node: %q is not allowed to modify pods %q annotations", nodeName, newPod.Name)
+	}
+
+	changes := mapDiff(oldPod.Annotations, newPod.Annotations)
+	changedKeys := maps.Keys(changes)
+
+	// ovnkube-node is not allowed to change annotations outside of it's scope
+	if !p.annotationKeys.HasAll(changedKeys...) {
+		return nil, fmt.Errorf("ovnkube-node on node: %q is not allowed to set at least one of the annotations on pod: %q: %v", nodeName, newPod.Name, changedKeys)
+	}
+
+	for _, key := range changedKeys {
+		if check := p.annotations[key]; check != nil {
+			if err := check(p.nodeLister, changes[key], newPod, nodeName); err != nil {
```

Comment:
maybe your check needs to consider old pod annotation values as well, so you can compare to see if someone is trying to change the IP address in the annotation, which shouldn't be allowed.

Reply:
If the old pod annotation is also checked, then ovn-node might get stucked to update the pod. Let's imagine that user creates the pod with wrong annotation. Current implementation allows ovn to update the annotation value with right value, but if this webhook check old value, then webhook blocks to edit becuase of the old value check.

I just imagine that some user just create Pods, based on previous runnining pod (which contains old annotation value, that might be wrong)....

Reply:
I would like to keep the checks to the minimum required for ensuring that pods running on other nodes are not compromised. We could build on top of the existing changes in the future but considering the corner cases we can stumble across this seems like a good start to me.

Reply:
if someone creates a pod with wrong annotation, ovnk will respect that annotation and try to use the values from it. It wont assign it a new ip address, because kapi and the annotation is considered the source of truth. @kyrtapz from what you told me this morning you were considering adding a check to make sure the annotation was not changing if it was already created. What are your plans here?

Reply:
Although now that I think about it, secondary network controllers might update the annotation with their network information as well after the primary network as been added...

Reply:
I don't think we should consider pod creation issue in this PR: https://github.com/ovn-org/ovn-kubernetes/pull/3830#discussion_r1318416051
Considering we are not handling that here is there a significant reason to block updates to annotations? Especially considering we might be missing some corner cases.

Reply:
yes lets ignore it for now and tackle it later